roadmap/parachains.md

```diff
@@ -0,0 +1,184 @@
+
+*Networking*: Protocols for nodes to speak to each other and transmit information across the network.
+
+*Node*: State or updates that must be maintained or performed by some or all nodes off-chain. Often interfaces with networking components, and references runtime state.
+
+---
+## Sub-projects and features:
+This section contains various sub-projects and the features that make them up.
+
+### Infrastructure/API
+
+#### *Peer Set Management*
+
+Category: Networking
+
+Validators assigned to a parachain need a way to discover and connect to collators in order to get fresh parachain blocks to validator.
+
+Collators need to discover and connect to validators in order to submit parachain blocks,
+
+Fishermen need to talk to validators and collators to fetch available data and circulate reports.
+
+Some connections are long-lived, some are ephemeral (just for a single request)
+
+#### Custom libp2p sub-protocols
+
+Polkadot parachains involve many distinct networking protocols. Ideally, we'd be able to spawn each of these as a separate futures task which communicates via channel with other protocols or node code as necessary. This requires changes in Substrate and libp2p.
+
+---
+### Assignment
+
+#### *Auctions*
+
+Category: Runtime
+
+Auctioning and registration of parachains. This is already implemented and follows the [Parachain Allocation — Research at W3F](https://research.web3.foundation/en/latest/polkadot/Parachain-Allocation.html) document.
+
+#### *Parathread Auctions*
+
+Category: Runtime
+
+Parathreads are pay-as-you-go parachains.  This consists of an on-chain mechanism for resolving an auction by collators and ensuring that they author a block.
+
+The node-side portion of parathreads is for collators to actually cast bids and to be configured for which conditions to cast bids under.
+
+#### *Validator Assignment*
+
+Category: Runtime
+
+Assignment of validators to parachains. Validators are only assigned to parachains for a short period of time. Tweakable parameters include length of time assigned to each parachain and length of time in advance that the network is aware of validators' assignments.
+
+---
+### Agreement
+
+#### *Attestation Circulation*
+
+Category: Networking
+
+A black-box networking component for circulating attestation messages (`Candidate`, `Valid`, `Invalid`) between validators of any given parachain to create a quorum on which blocks can be included.
+
+#### *Availability Erasure-coding*
+
+Category: Node, Networking
+
+For each potential, considered parachain block, perform and erasure-coding of the PoV and outgoing messages of the block. Call the number of validators on the relay chain for the Relay-chain block this parachain block is being considered for inclusion in `n`. Erasure-code into `n` pieces, where any `f + 1` can recover (`f` being the maximum number of tolerated faulty nodes = ~ `n / 3`). The `i'th` validator stores the `i'th` piece of the coding and provides it to any who ask.
```

Comment:
Shouldn't `f` be the minimum tolerated number of honest nodes, `n * 2 / 3 + 1`? If a node is honest, we can ask it for the chunk it's responsible for at any time (until finalty). So by asking every honest node we get enough pieces?

Reply:
There are two properties that we want - honest nodes must have their pieces and byzantine nodes can't prevent you from recovering by withholding their pieces.

I think every value from `f + 1` to `2f + 1` gives you both properties, as long as we can ensure that 1 honest node having its piece implies that other honest nodes have their pieces.

But using `f + 1` has a couple benefits over `2f + 1`:
  - Lower degree in the Reed-Solomon coding. This means we do much smaller polynomial interpolations (about 1/4th as much work since polynomial interpolation is O(N^2))
  - More combinations of honest nodes we could fully recover from. With `2f + 1`, if there are `f` byzantine nodes then you do have to reach _all_ the honest nodes. with `f + 1` redundancy you have `choose(f+1, 2f + 1)` combinations.

Reply:
(this is just the justification for the research - again, this document is not specifying anything new)